gnomad_qc/v4/sample_qc/generate_qc_mt.py

```diff
@@ -0,0 +1,293 @@
+    is_vds = isinstance(mtds, hl.vds.VariantDataset)
+    if test:
+        logger.info("Filtering to the first 20 partitions")
+        if is_vds:
+            mtds = hl.vds.VariantDataset(
+                mtds.reference_data._filter_partitions(range(20)),
+                mtds.variant_data._filter_partitions(range(20)),
+            )
+        else:
+            mtds = mtds._filter_partitions(range(20))
+
+    if is_vds:
+        vds = mtds
+    else:
+        logger.info("Converting MatrixTable to VariantDataset...")
+        mtds = mtds.select_entries("END", "LA", "LGT", "GQ", "DP", "LAD")
+        vds = hl.vds.VariantDataset.from_merged_representation(mtds)
+
+    if split:
+        logger.info("Performing split-multi...")
+        vds = hl.vds.split_multi(vds, filter_changed_loci=True)
+
+    logger.info("Filtering variants to predetermined QC variants...")
+    vds = hl.vds.filter_variants(vds, predetermined_qc_sites.ht())
+
+    logger.info("Densifying data...")
+    mt = hl.vds.to_dense_mt(vds)
+
+    logger.info("Writing dense MatrixTable...")
+    mt = mt.select_entries("GT", "GQ", "DP", "AD")
+
+    return mt
+
+
+def generate_qc_mt(
+    v3_mt: hl.MatrixTable,
+    v4_mt: hl.MatrixTable,
+    bi_allelic_only: bool = False,
+    min_af: float = 0.0001,
+    min_callrate: float = 0.99,
+    min_inbreeding_coeff_threshold: float = -0.8,
+    ld_r2: float = 0.1,
+) -> hl.MatrixTable:
+    """
+    Generate combined gnomAD v3 and v4 QC MatrixTable for use in relatedness and ancestry inference.
+
+    :param v3_mt: Dense gnomAD v3 MatrixTable filtered to predetermined sites.
+    :param v4_mt: Dense gnomAD v4 MatrixTable filtered to predetermined sites.
+    :param bi_allelic_only: Whether to filter to bi-allelic variants.
+    :param min_af: Minimum variant allele frequency to retain variant in QC MatrixTable.
+    :param min_callrate: Minimum variant callrate to retain variant in QC MatrixTable.
+    :param min_inbreeding_coeff_threshold: Minimum site inbreeding coefficient to retain variant in QC MatrixTable.
+    :param ld_r2: LD-pruning cutoff.
+    :return: MatrixTable of sites that pass QC filters.
+    """
+    logger.info(
+        "Number of (variants, samples) in the v3.1 MatrixTable: %s...", v3_mt.count()
+    )
+    logger.info(
+        "Number of (variants, samples) in the v4 MatrixTable: %s...", v4_mt.count()
+    )
+    # Remove v4 hard filtered samples
+    v4_mt = v4_mt.anti_join_cols(hard_filtered_samples.ht())
+
+    samples_in_both = v4_mt.cols().semi_join(v3_mt.cols())
+    n_samples_in_both = samples_in_both.count()
+    if n_samples_in_both > 0:
+        logger.info("The following samples are found in both MatrixTables:")
+        samples_in_both.show(n_samples_in_both)
+        raise ValueError("Overlapping sample names were found in the MatrixTables!")
+
+    logger.info(
+        "Performing a union of the v3.1 and v4 predetermined QC site MatrixTable columns..."
+    )
+    mt = v3_mt.union_cols(v4_mt)
```

Comment:
I will do a row comparison as a validity check, but yeah we shouldn't lose them due to v3, there could be some lost because of v4, but it should be minimal at this step